### DISCUSSION TITLE
DNS settings failed

### DISCUSSION THREAD

**Discussion by f18911238888** [sinamics/ztnet#229] [Q&A]
Hello, I followed the help to set up the DNS service, but multiple tests were unsuccessful. Could you please tell me the detailed setup steps? Thank you for your support.

 翻译效果如何

**Comment by f18911238888** [comment]
To build the zeronsd service, I need to apply for an API access token. It seems that this feature is not available in the ZT controller. Please let me know where it is. Thank you.

**Comment by sinamics** [comment]
ZeroNSD is currently incompatible with ZTNet, as integration between these two services has not been implemented yet.

**Reply by PatCas71** [reply]
Hi, :) Is this feature on the road map?

**Reply by sinamics** [reply]
This feature will be baked into the 2.0 release of Zerotier, although the exact release date is currently uncertain.
https://github.com/zerotier/zeronsd/blob/main/docs/quickstart.md#status

I can have another go with it, and see if we could create something similar when time permits.

**Reply by raku-cat** [reply]
Hi @sinamics, is there any updates on this? Bumped in to this today migrating a network over from the my.zerotier instance, realized it was probably a dumb assumption that the API would be identical :p. 
I note there was https://github.com/sinamics/ztnet/issues/207 and a PR mentioned in the issue as well, both of which are now closed, in the issue you link to page https://ztnet.network/usage/create_dns_host as the solution, but it's not quite an adequate replacement of zeronsd which functioned as a proper nameserver, users only had to enable Allow DNS on their clients, this is much preferred for me as most of my users are quite low tech!
Thanks in advance! Awesome tool :)

**Reply by sinamics** [reply]
My understanding was that something like this would be baked into Zerotier v2, so I’ve been holding off to see how that unfolds. That said, I’ll do some digging when time allows to see what’s feasible on the ZTNet side.

**Reply by raku-cat** [reply]
> My understanding was that something like this would be baked into Zerotier v2, so I’ve been holding off to see how that unfolds. That said, I’ll do some digging when time allows to see what’s feasible on the ZTNet side.

I think it's had that note about being integrated in to v2 for some years at this point haha, time will tell I suppose. In the mean time I've also began working on a solution just using hickory-dns directly since I believe that's what zeronsd uses (trust-dns currently, but the project renamed). Working on some modifications of the zt2hosts script to make a zone file, can update if I get something a little streamlined put together :)

**Reply by raku-cat** [reply]
Alright, I was able to get something working, it's not perfect, but it's easy enough to set up and then just runs in the background.<br><br>
I have a [python script](https://gist.github.com/raku-cat/e14d6c1d4e8a0665373f0486bc909c7d) using the existing zt2hosts bash script as a base to create a hickory-dns compatible zone file.
I'll slap a big ol' warning here at the front just as is in the description of the gist, a lot of this script was machine generated with AI, though I've reviewed the code and done a bit of testing, I can offer no guarantees it wont start thermonuclear war over your tic-tac-toe game :)   

It uses the same env variables as the original script, but only takes a network id, and relies on the domain being set in the DNS section on the network already.

It tries to be pretty robust about retaining fields it doesn't have to edit, it'll try and retain the contents of an existing zone file it thinks is unmanaged by it if possible, requiring a `--force` flag to even run at all in that case, and has a dry-run option to show what would be written at any time. 
What it does write out is basically as minimal as possible of a zone file as needed, and it wraps what it writes in a comment fenced area so the user can add their own custom records to the file, and those records should be retained as the file is modified by the script. 

The script will enumerate all named nodes, and create A records for each one, but will exclude any with no name, or those which are suffixed with `-nodns` in the name.

It restarts the `hickory-dns` systemd service any time it actually writes a zone file, either new or updating, and will soft-fail if it can't import dbus or the restarting goes wrong, and the systemd usage can be outright disabled with a flag `--no-sytemd`
The script will also give different exit codes based on results, exit code 0 is only for when a file is actually updated or written, exit 1 is if there's any errors such as wrong network id or permissions issues, and exit 3 is used when nothing has errored out but nothing was updated.

The original script also was only usable on personal networks, this one will actually search your networks list for the network, enumerate the organizations you have access to, then search the network lists of those for the network, so it works with both and without an organization ID.

So, the usage flow of this script is as follows, assuming a default setup of hickory-dns with its default loations:
1. Edit your network's DNS settings to have a Search Domain and one Server Address which is the zerotier IP of the host which will be running hickory-dns, the script only reads the first server ip, so configure accordingly.
2. On your DNS host, create a config toml for hickory-dns (default location is `/etc/named.toml`) containing at least the following:
```
listen_addrs_ipv4 = ["your.zt.host.ip"]

[[zones]]
zone = "my.ztnet.domain."
zone_type = "Primary"
file = "my.ztnet.domain.zone"
```
- Note that the `file` option must be named exactly with the zone name and end in `.zone` as that's how the script will name the file.
- Multiple `[[zones]]` sections can be added for multiple networks, just ensure the server listens on all the correct addresses.
3. Run the script at least once, assuming env variables are set (described in `--help`), the command would be `ztnet-dns.py <network id>`, 
- this will write a zone file in `/var/named` which is the default path hickory-dns will check for them, and will probably need root.
4. Start hickory-dns on the host

All clients with allowDNS should now be able to resolve `$nodename.$zone` using your hickory-dns server!

In order to make this dynamic, I use the following systemd units:
#### `ztnet-dns@.service`
```Desktop
[Unit]
Description=Update Hickory-DNS zone file for given zerotier network id %i

[Service]
Type=oneshot
ExecStart=/usr/bin/python3 /usr/local/bin/ztnet-dns.py %i
SuccessExitStatus=0 3
```
#### `ztnet-dns@.service.d/env.conf`
```Desktop
[Service]
Environment="ZTNET_API_TOKEN=1234567890abcdefghijkyourtokenhere"
Environment="API_ADDRESS=https://your.ztnet.instance"
```
#### `ztnet-dns@.timer`
```Desktop
[Unit]
Description=Check for zone updates for %i every minute
Requires=ztnet-dns@%i.servic

[Timer]
OnUnitActiveSec=60s
AccuracySec=1s
Unit=ztnet-dns@%i.service

[Install]
WantedBy=timers.target
```

With the above, you can then activate `systemctl enable --now ztnet-dns@<your networkid>.timer` and it should check for updates every minute :) 
You can also of course manually trigger the service with `systectl start ztnet-dns@<your networkid>.service` as needed, but other than that it'll basically run on its own, node removals, rnames, ip changes, removal of names, exclusion by `-nodns` suffix, all will be propagated every minute, and you can add your own custom records to the zone file after the managed block end comment and they should be retained.

This could probably all be done with dnsmasq or something similar, but as I said hickory-dns/trust-dns(old name) is what zeronsd used (though through the rust library, don't know rust unfortunately), and I was feeling adventurous, and hickory wasn't too bad to install or configure on arch at least.
